### ISSUE TITLE
[Feature Request] Format Strings or `f'string'`

### ISSUE THREAD

**Issue by amqndin** [modular/modular#398]
### Bug description

```shell
https://i.imgur.com/T4HF4dY.png
as you can see in the screenshot i was not able to use f'string' in mojo.
```

### Steps to reproduce

```shell
- bro = 'mojo'
print(f'hello, {bro}!')

- https://i.imgur.com/T4HF4dY.png

- Include anything else that might help us debug the issue.
```

### System information

```shell
%%python
import subprocess
subprocess.Popen(["cat", "/proc/self/cgroup"])
```

**Comment by Mogball** [comment]
Thanks for filing this issue. Mojo is missing this feature at the moment but it's something we plan to add.

**Comment by Kratosix** [comment]
asked in 
https://github.com/modularml/mojo/issues/86#issuecomment-1542587207

**Comment by svihar** [comment]
It would be great if the f-strings are implemented similar to how it is done in Scala: https://docs.scala-lang.org/scala3/book/string-interpolation.html

Basically, instead of hard-coded handling in the compiler, please provide a generic solution, where we can define our own interpolations. This allows for typesafe SQL, and other uses.

**Comment by ksandvik** [comment]
Most likely there will be f-strings as defined in Python for compatibility, and something else then that could be very different, maybe as an open-source initiative.

**Comment by svihar** [comment]
I don't think it need to be mutually exclusive. AFAIK, Scala is able to achieve f-string like functionality using a generic solution. The official f-function/macro(?) can be part of the SDK, while the community could provide other functionalities such as type-safe SQL.  My main request is to avoid implementing f-strings hard-coded into the compiler.

**Comment by drunkwcodes** [comment]
I would like f-string to be a language feature.
I use `print(f"{variable = }")` to debug all of the time.
It's so useful to make additional importing bothersome.

**Comment by svihar** [comment]
Some languages like Java have default imports for core language libraries (e.g. java.lang). f-strings can be such an automatic import, while still keeping the implementation as a library instead of in the compiler.

**Comment by lattner** [comment]
FWIW, I agree with the above - we can provide something more generic which is nonetheless compatible with Python.  Swift also has a bunch of user-extensible string interpolation features as well.

**Comment by lattner** [comment]
We probably won't wait this long, but ideally you design the parsing and unparsing components of a language together - eg you want to add enums (aka algebraic datatypes) when you add pattern matching support.  In this case, it means designing string interpolation and regex support together.  This ensures the concepts line up nicely.

Swift didn't do this, but did eventually add regex's.  I'm not arguing that Mojo should follow it's path, but if you're curious there are lots of links online about how it works: https://www.hackingwithswift.com/swift/5.7/regexes

**Comment by drunkwcodes** [comment]
[Swift's regex](https://developer.apple.com/documentation/swift/regex) is interesting that it uses `//` to quote to regex pattern.
[julia;s regex](https://docs.julialang.org/en/v1/manual/strings/#man-regex-literals) integrates with strings more. It even bakes the regex class into raw strings. I'm wondering if mojo can do this and be compatible with python at the same time.

The ruff linter story is astonishing:

![](https://user-images.githubusercontent.com/1309177/232603516-4fb4892d-585c-4b20-b810-3db9161831e4.svg)

We should make it in mojo too!

The regex engine behind ruff is rust regex, which provides a clear [benchmark](https://rust-leipzig.github.io/regex/2017/03/28/comparison-of-regex-engines/).

The engine which can defeat rust regex is `PCRE-JIT`.
I'm willing to try to make a wrapper for mojo.
Is it ok for us to choose `PCRE-JIT` as our regex engine?

It is a good test-run for my project manager.
Is there any brief tutorial for writing a C/C++ extension for mojo?

**Comment by iFrozenPhoenix** [comment]
Hello, are there any updates on this topic? Is this something that will be implemented in the short time?

**Comment by bethebunny** [comment]
FYI this was actually added some time ago. There's plenty of improvement to be made, but you can format strings via `String("{}, {}, {}").format(True, 4, "hello")`.

**Comment by arthurevans** [comment]
To follow up on what @bethebunny posted: the currently supported subset of `String.format()` is  documented here: https://docs.modular.com/mojo/stdlib/collections/string/format/ and this API doc page will be updated as new support is added.

**Comment by vexvec** [comment]
Wouldn't it be more intuitive to make the format function a method of the string type instead of a module function so that we can write "Hello {}".format("world")?

Idk but mixing in functional programming paradigms into a language that is heavily object orientated feels strange.

**Comment by mon-jai** [comment]
Also dedented string too.

It would be better if multi-line strings are dedented by default. Carbon take this path too.

**Comment by YkdWaWEzVmphR1Z1** [comment]
I know it's difficult to implement because f-strings can contain arbitrary expressions, not just identifiers. In fact, even Python did not properly implement them until version 3.12.

I consider f-strings to be iconic to Python; they should probably be implemented shortly after version 1.0.

Regarding efficiency, I propose having an FStringLiteral that coerces to a string and a trait like Stringable or overloads.

For example, print could simply write the f-string directly to a buffer without affecting usability or compatibility.

**Comment by BradLarson** [comment]
As of the 26.2 release, [Mojo now supports t-strings](https://docs.modular.com/mojo/std/format/tstring/TString/) as a generalization of f-strings (in the style of [PEP 750](https://peps.python.org/pep-0750/)):

```mojo
def main():
    var x = 10
    var y = 20
    print(t"{x} + {y} = {x + y}")  # prints: 10 + 20 = 30
```
